### PR TITLE
New Guide post: Integration with DynamoDB(Walk-throughs)

### PR DESCRIPTION
**Pull request by HarisaranG**
## Change Summary

- [x] Modify SVG
- [x] Add warnings
- [x] Change table name
- [x] Combine code snippets
- [x] Add AWS Docs link
## PR Checklist

- [x] I have read and signed the [Contributor License Agreement](https://forms.gle/PZyiY5N2GDQU8GsV9).

### CODE REVIEW

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 1]
```diff
@@ -0,0 +1,297 @@
+# DynamoDB Integration with Typesense
+
+Hey there! This post give you a guide to integrate Typesense custer with AWS DynamoDB by setting up a trigger with AWS Lambda.
+
+![Typesense DynamoDB Integration Chart](~@guide-images/typesense-dynamodb/dynamodb-typesense-integration-1.png)
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:5]: Could you use SVG logos and then export the file as an SVG as well? The current one looks quite blurry:

Screen Shot 2021-03-10 at 8 24 55 PM

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+## Create AWS DynamboDB with streams enabled
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:7]: Could you add Step numbers and make the titles shorter in general. 

See this article for an example: https://new-site.typesense.org/docs/0.19.0/guide/typesense-firebase.html

For eg: this particular title could be shortened to "Step 1: Create a DynamoDB table"

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+Create a DynamoDB with your choice of name and partition key ("id" is recommended). Now, after creating a DynamoDB database you should enable streams in the "Overview" section.
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:9]: > Create a DynamoDB
> after creating a DynamoDB database

DynamoDB is the name of the service. So you'd create a DynamoDB _table_

> you should enable streams

Could be phrased as "You want to enable streams"

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+![DynamoDB streams](~@guide-images/typesense-dynamodb/dynamodb-typesense-integration-2.png)
+
+If you are doing this in AWS CLI,
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:13]: > If you are doing this in AWS CLI,

You can also do this using the AWS CLI:

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+```bash
+aws dynamodb create-table \
+    --table-name typensense \
+    --attribute-definitions AttributeName=id,AttributeType=N \
+    --key-schema AttributeName=id,KeyType=HASH  \
+    --provisioned-throughput ReadCapacityUnits=5,WriteCapacityUnits=5 \
+    --stream-specification StreamEnabled=true,StreamViewType=NEW_AND_OLD_IMAGES
+```
+## Setting up Triggers for DyanmoDB
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:23]: Could you add Step number, and use an action verb. 

For eg: `Step 2:  Set up a trigger`

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+Create a trigger for DynamoDB database and this will lead you to AWS Lambada function's page from there you can select an existing funciton or a new function. In this tutorial the function name is ```typesense-indexing```
+
+![DynamoDB Triggers](~@guide-images/typesense-dynamodb/dynamodb-typesense-integration-3.png)
+
+Then now you need to create a "Lambda Execution Role" i.e give permission for your function. Now, head over to IAM Roles section and create a new role with three main permissions:
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:29]: > Then now you need to create a

Now let's create a...

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+
+* AmazonDynamoDBFullAccess
+* AmazonDynamoDBFullAccesswithDataPipeline
+* AWSLambdaBasicExecutionRole
+
+![IAM Execution Role](~@guide-images/typesense-dynamodb/dynamodb-typesense-integration-4.png)
+
+Now, attach this execution role to the trigger Lambda funtion of DynamoDB.
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:37]: Let's add a warning here saying: These IAM permissions are just examples for the purposes of this guide. In a production setting, please consult the IAM documentation to only grant the minimal permissions needed for your particular use case.

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+            ],
+            "Resource": "arn:aws:dynamodb:region:accountID:table/typesense/stream/*"
+        },
+    ]
+}
+```
+
+The policy has three statements that allow TypesenseLambdaRole to do the following:
+
+* Run a Lambda function ```typesense-indexing```. You create the function later in this tutorial.
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:102]: > You create the function later

We'll be creating the function later...

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+* Access Amazon CloudWatch Logs. The Lambda function writes diagnostics to CloudWatch Logs at runtime.
+* Read data from the DynamoDB stream for ```typesense```.
+
+Now, we are going to attach the above roles to our IAM execution role which we have created
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:106]: Now, let's attach the above roles...

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+
+```bash
+aws iam put-role-policy --role-name TypesenseLambdaRole \
+    --policy-name TypesenseLambdaRolePolicy \
+    --policy-document file://role-policy.json
+```
+
+## Importing ```typesense``` python package in Lambda function
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:114]: Since using Python with Lambda requires additional setup work, it takes focus away from the primary purpose of this article. Could we instead use Javascript here, which doesn't require the additional setup steps?

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+    'connection_timeout_seconds': 2
+})
+```
+Now, we need to create a collection. For creating a collection, check out the API Docs of typesense
+
+[Typesense API Docs](https://new-site.typesense.org/docs/0.19.0/api/collections.html#create-a-collection)
+
+When a data is added in DynamoDB, the Lambda function is called with `event` and `context` parameters. A sample event parameter form DynamoDB would be,
+
+```json
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:198]: Use `javascript` since json doesn't support inline comments

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+            "awsRegion": "us-east-2",
+            "eventName": "MODIFY", // this can 'INSERT', 'MODIFY' and 'DELETE'
+            "eventSourceARN": "<AWS-ARN>",
+            "eventSource": "aws:dynamodb"
+        },
+    ]
+}
+```
+
+A sample to code to to process above `event` response
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:225]: Could you rephrase this?

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+    --starting-position TRIM_HORIZON
+  ```
+* Make sure to test the function with a set of sample events before enabling the trigger.
+
+That's a wrap! Now your DynamoDB database will be automatically indexed in your Typesense cluster.
+
+## References
+- [Sample Code](https://github.com/HarisaranG/dynamodb-typesense-indexing)
+- [DynamoDB Streams](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.Lambda.Tutorial.html)
+- [Typesense API](https://typesense.org/docs/0.19.0/api/)
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:296]: Instead of hard-coding the version number, could you use `{{ $page.typesenseVersion }}`

**Review by jasonbosco — CHANGES_REQUESTED** [review]

File: new-site/docs-site/content/.vuepress/public/images/typesense-dynamodb/typesense-dynamodb.svg [hunk 14]
```diff
@@ -0,0 +1,77 @@
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="1920" height="1080" viewBox="0 0 1920 1080">
```
Comment (jasonbosco) [inline new-site/docs-site/content/.vuepress/public/images/typesense-dynamodb/typesense-dynamodb.svg:1]: Just noticed this... Could you remove the Python logo from this image? Makes it a little confusing.

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 15]
```diff
@@ -0,0 +1,240 @@
+# DynamoDB Integration with Typesense
+
+Hey there! This post give you a guide to integrate Typesense cluster with AWS DynamoDB by setting up a trigger with AWS Lambda.
+
+![Typesense DynamoDB Integration Chart](~@guide-images/typesense-dynamodb/typesense-dynamodb.svg)
+
+## Step 1: Create Typesense Cluster
+
+Sign up for a new account in [Typesense Cloud](https://cloud.typesense.org/) and get Endpoint URL, Port number and API key.
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:9]: After that line... 

We're using Typesense Cloud for this walk-through since we need a public Typesense endpoint for the Lambda function to be able write to. 

You can also self-host Typesense on a server of your choice. See this (link to Installing Typesense section) for more details on how to self-host Typesense.

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+## Step 2: Create a DynamoDB table
+
+Create a DynamoDB table with your choice of name and partition key ("id" is recommended).
+
+After creating a DynamoDB table you want to enable streams in the "Overview" section.
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:14]: ... Overview section of the AWS console.

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+
+You can also do this using AWS CLI:
+
+```bash
+aws dynamodb create-table \
+    --table-name typensense \
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:20]: Could you change the table name to something else? May be `YourTableName` or something

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+    --attribute-definitions AttributeName=id,AttributeType=N \
+    --key-schema AttributeName=id,KeyType=HASH  \
+    --provisioned-throughput ReadCapacityUnits=5,WriteCapacityUnits=5 \
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:23]: Is `--provisioned-throughput ReadCapacityUnits=5,WriteCapacityUnits=5` a required parameter? 

Could we only have the absolute required parameters, in order to make it as simple as possible?

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+
+## Step 3: Create Lambda Execution Role
+
+Now let's create a "Lambda Execution Role" i.e give permission for your function, head over to IAM Roles section and create a new role with three main permissions:
+
+* AmazonDynamoDBFullAccess
+* AmazonDynamoDBFullAccesswithDataPipeline
+* AWSLambdaBasicExecutionRole
+
+These IAM role permissions are just examples for the purposes of this guide. Before deploying for production, please consult the IAM documentation to only grant the minimal permissions needed for your particular use case.
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:35]: Surround this block with 

```
:::warning
These IAM role permissions...
:::
```

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+Now, we are going to attach the above roles to our IAM execution role which we have created
+
+```bash
+aws iam put-role-policy --role-name TypesenseLambdaRole \
+    --policy-name TypesenseLambdaRolePolicy \
+    --policy-document file://role-policy.json
+```
+## Step 4: Create a Lambda Function
+
+Head over to Lambda section of AWS and create a new Lambda function with the above created execution role.
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:112]: Lambda section of the AWS *console*.

Could you also add a link to the AWS documentation on how to do this, in case people need a reminder?

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+
+Now, let's create a new python script file named `lambda_function.py`. For creating collections and indexing into your cloud cluster refer: [Typesense API](../api/README.md)
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:114]: See the *API documentation* for detailed information about all the parameters available to create collections and documents. 

Also, let's move this sentence below the code example.

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+
+```python
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:116]: Could you combine this snippet with the next one, so the entire code snippet that a user would write inside the lamdba is in one place?

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+    else:
+        upload = ddb_record['NewImage'] # format your document here and the use upsert function to index it.
+        res = client.collections['<collection-name>'].upsert(upload)
+        print(res)
+    processed = processed + 1
+
+print('Successfully processed {} records'.format(processed))
+```
+
+Note: Install all your code using `pip install <dependency-name> -t .`. This will install all the dependencies for the function in the current directory.
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:149]: Install all your dependencies using...

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+    --handler lambda_function.lambda_handler \
+    --timeout 5 \
+    --runtime python3.7
+    ```
+
+Make sure to test the function with a set of sample events before enabling the trigger.
+
+## Step 5: Setup up a trigger
+
+Now, navigate to DynamoDB table &#8594; Triggers and add this existing Lambda function to that table.
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:183]: Now, navigate to your DynamoDB table in the AWS Console, then visit the Triggers section...

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 25]
```diff
@@ -0,0 +1,240 @@
...
+
+A example `event` response from DynamoDB table,
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:185]: This code snippet looks out of place here. Might be better to move it under the previous step and say here's what an example event that we'll process using the lambda looks like

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md [hunk 26]
```diff
@@ -0,0 +1,240 @@
...
+            "eventSource": "aws:dynamodb"
+        },
+    ]
+}
+```
+You can also do this using the AWS CLI:
+
+* Get ARN for DynamoDB table
+    ```bash
+    aws dynamodb describe-table --table-name `table-name`
```
Comment (jasonbosco) [inline new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:216]: Use same table name from the example above

**Review by jasonbosco — CHANGES_REQUESTED** [review]

File: new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md (hunk 18 above)
Comment (HarisaranG) [reply new-site/docs-site/content/0.19.0/guide/typesense-dynamodb.md:23]: If we use that one we can stay on free tier for a long period this is the minimum configuration. But it is not necessary.

**Review by jasonbosco — APPROVED** [review]
